src/main/java/org/radarcns/listener/managementportal/ManagementPortalClient.java

```diff
@@ -0,0 +1,266 @@
+            List<Subject> allSubjects = mapper.readValue(response.body().string(), mapper
+                    .getTypeFactory().constructCollectionType(List.class, Subject.class));
+            logger.info("Retrieved Subjects from MP from Project " + projectName);
+            return allSubjects;
+        } else if (response.code() == HTTP_NOT_FOUND) {
+            logger.info("Couldn't get any subjects for project :", projectName);
+            throw new NotFoundException("Cannot find any subjects under project : " + projectName
+                    + "No subjects registered under this project or invalid projectName");
+        }
+        return null;
+    }
+
+    /**
+     * Retrieves all {@link Project} from Management Portal using {@link ServletContext} entity.
+     *
+     * @return {@link ArrayList} of {@link Project} retrieved from the Management Portal
+     */
+    public List<Project> getAllProjects() throws
+            IOException {
+        ManagementPortalConfig config = Properties.getApiConfig().getManagementPortalConfig();
+        URL getAllProjectsUrl = new URL(config.getManagementPortalUrl(),
+                config.getProjectEndpoint());
+        Request getAllProjects = this.buildGetRequest(getAllProjectsUrl);
+        Response response = this.client.newCall(getAllProjects).execute();
+        if (response.isSuccessful()) {
+            List<Project> allProjects = mapper.readValue(response.body().string(), mapper
+                    .getTypeFactory().constructCollectionType(List.class, Project.class));
+            logger.info("Retrieved Projects from MP");
+            return allProjects;
+        }
+        return null;
+    }
+
+    /**
+     * Retrieves a {@link Project} from the Management Portal using {@link ServletContext} entity.
+     *
+     * @param projectName {@link String} of the Project that has to be retrieved
+     * @return {@link Project} retrieved from the Management Portal
+     */
+    public Project getProject(String projectName) throws IOException, NotFoundException {
+
+        ManagementPortalConfig config = Properties.getApiConfig().getManagementPortalConfig();
+        URL getProjectFromProjectName = new URL(config.getManagementPortalUrl(),
+                config.getProjectEndpoint() + '/' + projectName);
+        Request getProject = this.buildGetRequest(getProjectFromProjectName);
+        Response response = this.client.newCall(getProject).execute();
+        if (response.isSuccessful()) {
+            Project project = mapper.readValue(response.body().string(), Project.class);
+            logger.info("Retrieved project {} from MP", projectName);
+            return project;
+        } else if (response.code() == HTTP_NOT_FOUND) {
+            logger.info("Couldn't get project details for project :", projectName);
+            throw new NotFoundException("Cannot project data for project-name : " + projectName
+                    + " Invalid projectName");
+        }
+        return null;
```

Comment:
Should throw an exception here.

Reply:
Indeed. There is still a lot of room for improvements. E.g. there is a local list of subjects which was used as cache. But it would introduce inconsistencies. I don't use them, but it is not completely removed. I think if we need a cache there, it should be more systematic.  Also i agree the exception handling can be improved.